pkg/loki/loki.go

```diff
@@ -159,72 +170,145 @@ func (t *Loki) setupAuthMiddleware() {
+			}
 		}
+
+		level.Error(util.Logger).Log("msg", "module failed", "module", "unknown", "error", service.FailureCase())
+	}
+
+	sm.AddListener(services.NewManagerListener(healthy, stopped, serviceFailed))
+
+	// Currently it's the Server that reacts on signal handler,
+	// so get Server service, and wait until it gets to Stopping state.
+	// It will also be stopped via service manager if any service fails (see attached service listener)
+	// Attach listener before starting services, or we may miss the notification.
+	serverStopping := make(chan struct{})
+	t.serviceMap[Server].AddListener(services.NewListener(nil, nil, func(from services.State) {
+		close(serverStopping)
+	}, nil, nil))
+
+	// Start all services. This can really only fail if some service is already
+	// in other state than New, which should not be the case.
+	err = sm.StartAsync(context.Background())
+	if err == nil {
+		// no error starting the services, now let's just wait until Server module
+		// transitions to Stopping (after SIGTERM or when some service fails),
+		// and then initiate shutdown
+		<-serverStopping
 	}
-}
 
-func (t *Loki) stopping(m moduleName) {
-	t.stoppingModule(m)
-	deps := orderedDeps(m)
-	// iterate over our deps in reverse order and call stoppingModule
-	for i := len(deps) - 1; i >= 0; i-- {
-		t.stoppingModule(deps[i])
+	// Stop all the services, and wait until they are all done.
+	// We don't care about this error, as it cannot really fail.
+	_ = services.StopManagerAndAwaitStopped(context.Background(), sm)
+
+	// if any service failed, report that as an error to caller
+	if err == nil {
+		if failed := sm.ServicesByState()[services.Failed]; len(failed) > 0 {
+			for _, f := range failed {
+				if f.FailureCase() != util.ErrStopProcess {
+					// Details were reported via failure listener before
+					err = errors.New("failed services")
+					break
+				}
+			}
+		}
 	}
+	return err
 }
 
-func (t *Loki) stoppingModule(m moduleName) {
-	level.Info(util.Logger).Log("msg", "notifying module about stopping", "module", m)
-	if modules[m].stopping != nil {
-		if err := modules[m].stopping(t); err != nil {
-			level.Error(util.Logger).Log("msg", "error stopping", "module", m, "err", err)
+func (t *Loki) readyHandler(sm *services.Manager) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if !sm.IsHealthy() {
+			msg := bytes.Buffer{}
+			msg.WriteString("Some services are not Running:\n")
+
+			byState := sm.ServicesByState()
+			for st, ls := range byState {
+				msg.WriteString(fmt.Sprintf("%v: %d\n", st, len(ls)))
+			}
+
+			http.Error(w, msg.String(), http.StatusServiceUnavailable)
+			return
+		}
+
+		// Ingester has a special check that makes sure that it was able to register into the ring,
+		// and that all other ring entries are OK too.
+		if t.ingester != nil {
+			if err := t.ingester.CheckReady(r.Context()); err != nil {
+				http.Error(w, "Ingester not ready: "+err.Error(), http.StatusServiceUnavailable)
```

Comment:
should you `return` after here ?

Reply:
Yes! (Already fixed in Cortex, but forgot to fix here)